erpnext/stock/stock_ledger.py

```diff
@@ -241,46 +258,64 @@ def get_serialized_values(self, sle):
+					sle.voucher_type, sle.voucher_no, sle.batch_no, self.allow_zero_rate,
+					currency=erpnext.get_company_currency(sle.company), batch_wise_valuation=0)
 
 	def get_moving_average_values(self, sle):
 		actual_qty = flt(sle.actual_qty)
-		new_stock_qty = flt(self.qty_after_transaction) + actual_qty
-		if new_stock_qty >= 0:
+		prev_qty = flt(self.batch_data.batch_qty_after_transaction if self.batch_wise_valuation else self.qty_after_transaction)
+		new_qty = prev_qty + actual_qty
+		prev_valuation_rate = flt(self.batch_data.batch_valuation_rate if self.batch_wise_valuation else self.valuation_rate)
+		new_valuation_rate = prev_valuation_rate
+
+		if new_qty >= 0:
 			if actual_qty > 0:
-				if flt(self.qty_after_transaction) <= 0:
-					self.valuation_rate = sle.incoming_rate
+				if prev_qty <= 0:
+					new_valuation_rate = sle.incoming_rate
 				else:
-					new_stock_value = (self.qty_after_transaction * self.valuation_rate) + \
+					new_stock_value = (prev_qty * prev_valuation_rate) + \
 						(actual_qty * sle.incoming_rate)
 
-					self.valuation_rate = new_stock_value / new_stock_qty
+					new_valuation_rate = new_stock_value / new_qty
 
 			elif sle.outgoing_rate:
-				if new_stock_qty:
-					new_stock_value = (self.qty_after_transaction * self.valuation_rate) + \
+				if new_qty:
+					new_stock_value = (prev_qty * prev_valuation_rate) + \
 						(actual_qty * sle.outgoing_rate)
 
-					self.valuation_rate = new_stock_value / new_stock_qty
+					new_valuation_rate = new_stock_value / new_qty
 				else:
-					self.valuation_rate = sle.outgoing_rate
+					new_valuation_rate = sle.outgoing_rate
 
 		else:
-			if flt(self.qty_after_transaction) >= 0 and sle.outgoing_rate:
-				self.valuation_rate = sle.outgoing_rate
+			if prev_qty >= 0 and sle.outgoing_rate:
+				new_valuation_rate = sle.outgoing_rate
 
-			if not self.valuation_rate and actual_qty > 0:
-				self.valuation_rate = sle.incoming_rate
+			if not new_valuation_rate and actual_qty > 0:
+				new_valuation_rate = sle.incoming_rate
 
 			# Get valuation rate from previous SLE or Item master, if item does not have the
 			# allow zero valuration rate flag set
-			if not self.valuation_rate and sle.voucher_detail_no:
+			if not new_valuation_rate and sle.voucher_detail_no:
 				allow_zero_valuation_rate = self.check_if_allow_zero_valuation_rate(sle.voucher_type, sle.voucher_detail_no)
 				if not allow_zero_valuation_rate:
-					self.valuation_rate = get_valuation_rate(sle.item_code, sle.warehouse,
-						sle.voucher_type, sle.voucher_no, self.allow_zero_rate,
-						currency=erpnext.get_company_currency(sle.company))
+					new_valuation_rate = get_valuation_rate(sle.item_code, sle.warehouse,
+						sle.voucher_type, sle.voucher_no, sle.batch_no, self.allow_zero_rate,
+						currency=erpnext.get_company_currency(sle.company), batch_wise_valuation=self.batch_wise_valuation)
+
+		self.qty_after_transaction += flt(sle.actual_qty)
+		if self.batch_wise_valuation:
```

Comment:
move this to a smaller function?

Reply:
The calculation? The calculation is very specific to this function. It can be done, but I feel it'll make it more confusing since I'm already using `self.batch_data` in the beginning of the method.

Reply:
You will have to elaborate on this. This is not as simple as: move to new function. This would require some minor refactoring.